src/Rematch.jl

```diff
@@ -128,42 +128,76 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
+        # struct or extractor call
+        if !isempty(string(T)) && islowercase(first(string(T)))
+            # Extractor call.
+            result = gensym("unapply")
+            len = length(subpatterns)
+            # Extractor call.
+            # The function should take one argument and return either `nothing`, if the argument does not match,
+            # or a tuple if it does match.
+            if len == 0
+                # If there are no subpatterns, the result value is just checked for nothingness.
+                quote
+                    $(esc(T))($value) !== nothing
                 end
-                if !(isstructtype(typeof($(esc(T)))) || issabstracttype(typeof($(esc(T)))))
-                    throw(LoadError("Attempted to match on a pattern that is not a struct", @__LINE__, AssertionError("Incorrect match usage")))
+            elseif len == 1
+                # If there is just one subpattern, the result value is matched against it.
+                quote
+                    begin
+                        $result = $(esc(T))($value)
+                        $result !== nothing && $(handle_destruct(result, subpatterns[1], bound, asserts))
+                    end
                 end
-                # This assertion is necessary:
-                # If $expected_fieldcount < $actual_fieldcount, to catch missing fields.
-                # If $expected_fieldcount > $actual_fieldcount, to avoid a BoundsError.
-                $expected_fieldcount = evaluated_fieldcount($(esc(T)))
-                $actual_fieldcount = $(esc(len))
-                if $expected_fieldcount != $actual_fieldcount
-                    error("Pattern field count is $($actual_fieldcount) expected $($expected_fieldcount)")
+            else
+                # If there is more than one subpattern, the result value is matched against a tuple pattern.
+                quote
+                    begin
+                        $result = $(esc(T))($value)
+                        $result !== nothing &&
+                        ($result isa Tuple) &&
+                        $(handle_destruct_fields(result, pattern, subpatterns, :(length($result)), :getindex, bound, asserts; allow_splat=true))
+                    end
                 end
-                 
-            end)
+            end
         else
-            # Pattern uses named arguments to refer to fields e.g. Foo(x=0, y=true)
-            # Could assert that the expected field names are a subset of the actual field names.
-            # Can omit the assertion because if the field doesn't exist getfield() will fail with "type $T has no field $field".
-        end
-        quote
-            # I would prefer typeof($value) == $(esc(T)) but this doesn't convey type information in Julia 0.6
-            $value isa $(esc(T)) &&
-            $(handle_destruct_fields(value, pattern, subpatterns, length(subpatterns), :getfield, bound, asserts; allow_splat=false))
+            # Struct.
```

Comment:
The diff got chopped up in a really annoying way. I think you didn't change anything about the existing implementation for matching structs, right?

Reply:
no, just whitespace.

Reply:
@comnik you can disable whitespace in github's diff, i dunno if you've seen that? :)